java/ql/src/experimental/Security/CWE/CWE-208/NonConstantTimeCheckOnSignatureQuery.qll

```diff
@@ -0,0 +1,323 @@
+
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(ProduceCryptoCall call | call.getQualifier() = sink.asExpr())
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow2::Node fromNode, DataFlow2::Node toNode) {
+    updateCryptoOperationStep(fromNode, toNode)
+    or
+    createMessageDigestStep(fromNode, toNode)
+    or
+    updateMessageDigestStep(fromNode, toNode)
+  }
+}
+
+/** A source that produces result of cryptographic operation. */
+class CryptoOperationSource extends DataFlow::Node {
+  ProduceCryptoCall call;
+
+  CryptoOperationSource() { call.output() = this.asExpr() }
+
+  /** Holds if remote user input was used in the cryptographic operation. */
+  predicate includesUserInput() {
+    exists(
+      DataFlow2::PathNode source, DataFlow2::PathNode sink, UserInputInCryptoOperationConfig config
+    |
+      config.hasFlowPath(source, sink)
+    |
+      sink.getNode().asExpr() = call.getQualifier()
+    )
+  }
+
+  /** Gets a method call that produces cryptographic result. */
+  ProduceCryptoCall getCall() { result = call }
+}
+
+/** Methods that use a non-constant-time algorithm for comparing inputs. */
+private class NonConstantTimeEqualsCall extends MethodAccess {
+  NonConstantTimeEqualsCall() {
+    getMethod()
+        .hasQualifiedName("java.lang", "String", ["equals", "contentEquals", "equalsIgnoreCase"]) or
+    getMethod().hasQualifiedName("java.nio", "ByteBuffer", ["equals", "compareTo"])
+  }
+}
+
+/** A static method that uses a non-constant-time algorithm for comparing inputs. */
+private class NonConstantTimeComparisonCall extends StaticMethodAccess {
+  NonConstantTimeComparisonCall() {
+    getMethod().hasQualifiedName("java.util", "Arrays", ["equals", "deepEquals"]) or
+    getMethod().hasQualifiedName("java.util", "Objects", "deepEquals") or
+    getMethod()
+        .hasQualifiedName("org.apache.commons.lang3", "StringUtils",
+          ["equals", "equalsAny", "equalsAnyIgnoreCase", "equalsIgnoreCase"])
+  }
+}
+
+/**
+ * A config that tracks data flow from remote user input to methods
+ * that compare inputs using a non-constant-time algorithm.
+ */
+private class UserInputInComparisonConfig extends TaintTracking2::Configuration {
+  UserInputInComparisonConfig() { this = "UserInputInComparisonConfig" }
+
+  override predicate isSource(DataFlow::Node source) { source instanceof RemoteFlowSource }
+
+  override predicate isSink(DataFlow::Node sink) {
+    exists(NonConstantTimeEqualsCall call |
+      sink.asExpr() = [call.getAnArgument(), call.getQualifier()]
+    )
+    or
+    exists(NonConstantTimeComparisonCall call | sink.asExpr() = call.getAnArgument())
+  }
+}
+
+/** Holds if `expr` looks like a constant. */
+private predicate looksLikeConstant(Expr expr) {
+  expr.isCompileTimeConstant()
+  or
+  expr.(VarAccess).getVariable().isFinal() and expr.getType() instanceof TypeString
```

Comment:
> I assume this might result in false positives, for example some people tend to make all parameters and local variables final, but that does not mean that their value is a constant.

It is true for many types. But `final String` is effectively a constant because `String` is immutable. I think the line 235 would rather result in false negatives than false positives since `NonConstantTimeComparisonSink` calls `not looksLikeConstant(anotherParameter)`.

> Note that CompileTimeConstantExpr matches strings as well, and already covers read access of a variable which is final and whose initializer is a CompileTimeConstantExpr.

If I remember correctly, I added the line 235 to avoid false positives reported against OpenJDK. Something like the following:

```java
    public static class ConstantHolder {

        static final ConstantHolder A = new ConstantHolder("a");

        private final String constant;

        ConstantHolder(String constant) {
            this.constant = constant;
        }

        @Override
        public String toString() {
            return constant;
        }
    }

    public static final String CONSTANT = ConstantHolder.A.toString();

    public boolean test(byte[] data) throws Exception {
        Mac mac = Mac.getInstance("HmacSHA256");
        byte[] actualMac = mac.doFinal(data);
        return CONSTANT.equals(new String(actualMac));
    }
```

The query doesn't really check what `ConstantHolder.toString()` does but is satisfied with `final String CONSTANT`. That's why the method is called `looksLikeConstant`.

I'd prefer to keep the predicate as-is.

Reply:
> But `final String` is effectively a constant because `String` is immutable

Yes, but only if the `String` is a constant as well. Code which (for example due to the used code style) has all parameters `final` would not be detected by this query, e.g.:
```java
// Assuming `signature` is tainted
public boolean areSignaturesEqual(final String signature, final String expectedSignature) {
    return signature.equals(expectedSignature);
}
```

If you noticed false positives for fields, then maybe this queck should only be made for fields to reduce false negatives for parameters and local variables.

What do you think?